ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala

```diff
@@ -0,0 +1,221 @@
+}
+import com.daml.ledger.participant.state.kvutils.api.{LedgerReader, LedgerRecord, LedgerWriter}
+import com.daml.ledger.participant.state.kvutils.{Envelope, KeyValueCommitting}
+import com.daml.ledger.participant.state.v1._
+import com.digitalasset.daml.lf.data.Ref
+import com.digitalasset.daml.lf.data.Time.Timestamp
+import com.digitalasset.daml.lf.engine.Engine
+import com.digitalasset.ledger.api.health.{HealthStatus, Healthy}
+import com.digitalasset.platform.akkastreams.dispatcher.Dispatcher
+import com.digitalasset.platform.akkastreams.dispatcher.SubSource.RangeSource
+import com.google.protobuf.ByteString
+import com.zaxxer.hikari.{HikariConfig, HikariDataSource}
+import javax.sql.DataSource
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Random
+
+class SqlLedgerReaderWriter(
+    ledgerId: LedgerId = Ref.LedgerString.assertFromString(UUID.randomUUID.toString),
+    val participantId: ParticipantId,
+    queries: Queries,
+    connectionSource: DataSource with AutoCloseable,
+)(implicit executionContext: ExecutionContext, materializer: Materializer)
+    extends LedgerWriter
+    with LedgerReader
+    with AutoCloseable {
+
+  private val engine = Engine()
+
+  private val dispatcher: Dispatcher[Index] =
+    Dispatcher(
+      "sql-participant-state",
+      zeroIndex = FirstIndex,
+      headAtInitialization = FirstIndex,
+    )
+
+  private val randomNumberGenerator = new Random()
+
+  // TODO: implement
+  override def checkHealth(): HealthStatus = Healthy
+
+  override def close(): Unit = {
+    dispatcher.close()
+    connectionSource.close()
+  }
+
+  override def retrieveLedgerId(): LedgerId = ledgerId
+
+  override def events(offset: Option[Offset]): Source[LedgerRecord, NotUsed] =
+    dispatcher
+      .startingAt(
+        offset.getOrElse(FirstOffset).components.head,
+        RangeSource((start, end) =>
+          withConnection { implicit connection =>
+            Future(Source(queries.selectFromLog(start, end)))
+        })
+      )
+      .map { case (_, record) => record }
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
+    withConnection { implicit connection =>
+      Future {
+        val submission = Envelope
+          .openSubmission(envelope)
+          .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
+        val stateInputKeys: Set[DamlStateKey] = submission.getInputDamlStateList.asScala.toSet
+        val stateInputs = readState(stateInputKeys)
+        val entryId = allocateEntryId()
+        val (logEntry, stateUpdates) = KeyValueCommitting.processSubmission(
+          engine,
+          entryId,
+          currentRecordTime(),
+          LedgerReader.DefaultTimeModel,
+          submission,
+          participantId,
+          stateInputs,
+        )
+        verifyStateUpdatesAgainstPreDeclaredOutputs(stateUpdates, entryId, submission)
```

Comment:
This assertion should move into `processSubmission` here and removed from here and from the other places that have this copied (file, in-memory).

Reply:
Agreed, just wanted to discuss this with you and @miklos-da first.

Reply:
I have added a ticket to track this work in the FCL Jira.